Add a troubleshooting guide for common failures in CI services

praneetloke opened the pull request:
### Proposed changes

As the title states, this adds a new page under our Continuous Delivery category to help our users troubleshoot common CI/CD failures when running Pulumi. This guide will be linked to from the Pulumi Console as part of its workflow configuration wizard.

### Unreleased product version (optional)

N/A

### Related issues (optional)

N/A

jkisk commented on content/docs/guides/continuous-delivery/troubleshooting-guide.md:
```diff
@@ -0,0 +1,122 @@
+Create one [here](https://app.pulumi.com/account/tokens) by logging in with the appropriate account.
+* A stack that you would like to update the automated pipeline.
+* Build tools (more on this below) based on the runtime of your Pulumi app.
+* Restore all dependencies for your Pulumi app.
```
Perhaps rephrase this to match the rest of the list, which are all things rather than actions?

praneetloke replied:
Ah great catch! I'll drop "restore".

jkisk commented on content/docs/guides/continuous-delivery/troubleshooting-guide.md:
```diff
@@ -0,0 +1,122 @@
+
+* Ensure that the account represented by the token you are using, has access to the stack.
+  * This can lead to 404s being returned from the Pulumi Service because the token is invalid for any number of reasons.
+* Ensure that you use the fully-qualified stack name when passing the stack name to `pulumi` commands.
```
perhaps an example  of a fqsn here?

praneetloke replied:
Oh shoot! I totally meant to do that. Glad you caught this. :)

jkisk commented on content/docs/guides/continuous-delivery/troubleshooting-guide.md:
```diff
@@ -0,0 +1,122 @@
+* Ensure that you use the fully-qualified stack name when passing the stack name to `pulumi` commands.
+* Ensure that you are running the `pulumi` command from the folder containing the `Pulumi.yaml` file that contains your Pulumi project's name.
+* If your stack has configuration, then the stack configuration file must also be co-located with the project file.
+  * For example, for a stack named `production`, the `Pulumi.production.yaml` file must exist alongside the `Pulumi.production.yaml`.
```
do you mean alongside the pulumi.yaml?

praneetloke replied:
Oops. Yes, I did.

jkisk commented on content/docs/guides/continuous-delivery/troubleshooting-guide.md:
```diff
@@ -0,0 +1,122 @@
+
+## Build Tools
+
+Pulumi invokes the build tool that corresponds to your Pulumi project's runtime. If your pipeline is missing one of these tools,
```
perhaps examples here?

praneetloke replied:
Done

jkisk commented on content/docs/guides/continuous-delivery/troubleshooting-guide.md:
```diff
@@ -0,0 +1,122 @@
+* There is an exception to restoring dependencies automatically for `.NET` when you use a private package feed. You must ensure that the
+package(s) from the private feed are accessible or you can use a pre-built binary (see the `binary` option [here]({{< relref "/docs/intro/concepts/project" >}})) with Pulumi to avoid rebuilding your `.NET` solution again.
+  * Note that if you do choose to use a pre-built binary, automatic plugin acquisition for Pulumi
+* You are caching the library dependencies but not the Pulumi plugins. Some services offer dependency caching by captuing a specific folder and restoring
```
captuing > capturing

jkisk commented on content/docs/guides/continuous-delivery/troubleshooting-guide.md:
```diff
@@ -0,0 +1,122 @@
+If your error is not related to any of the above categories, it is likely that there is a problem with the code in your infrastructure code.
+Our Programming Model docs can be very helpful in understanding the concepts of programming an infrastructure app. If that doesn't help you solve your
+particular issue, then there is always the power of the strong, knowledgable [Pulumi Slack Community](https://slack.pulumi.com). Signup is free and quick. In addition to other
+community members, various members of the Pulumi team themsevles hang out in the Slack channels and are willing to help you.
```
themsevles > themselves
maybe "would be happy to help you" instead of "are willing"?

praneetloke replied:
Definitely!

jkisk reviewed (approved):
This looks incredibly helpful!

cnunciato commented on content/docs/guides/continuous-delivery/troubleshooting-guide.md:
```diff
@@ -0,0 +1,129 @@
+
+Another common point of failure is the cloud provider credentials. The class of errors related to the credentials can be, but not limited to:
+
+* Incorrect credentials (wrong account, mismatched keys etc.)
```
```suggestion
* Incorrect credentials (wrong account, mismatched keys, etc.)
```

cnunciato commented on content/docs/guides/continuous-delivery/troubleshooting-guide.md:
```diff
@@ -0,0 +1,129 @@
+
+### Tips
+
+* Cloud provider credentials not specified using a recognized mechanism. In almost all cases, specifying the credentials through environment variables will work.
```
This doesn't read like a tip. Should it start with a "Make sure" or the like? In general, these should probably all read in the second person: "Make sure you've exposed your provider credentials in a supported way", for example, and ideally with a link to what the supported ways are.

praneetloke replied:
I will rephrase this sentence to be similar to what you suggest.

cnunciato commented on content/docs/guides/continuous-delivery/troubleshooting-guide.md:
```diff
@@ -0,0 +1,129 @@
+* Build tools (more on this below) based on the runtime of your Pulumi app.
+* Dependencies for your Pulumi app.
+* The right cloud provider credentials.
+
```
Worth mentioning that Pulumi itself should be installed? Perhaps obvious, but could be a good opportunity to describe or link off to the proper ways of installing Pulumi in CI, mention that Pulumi comes pre-installed on GitHub Actions builders, etc.

praneetloke replied:
Yes, absolutely! Completely forgot that.

praneetloke replied:
I added an item here about the Pulumi CLI, as well as a section about it.

cnunciato commented on content/docs/guides/continuous-delivery/troubleshooting-guide.md:
```diff
@@ -0,0 +1,129 @@
+* Ensure that the Pulumi step has access to the environment variables containing the credentials.
+  * If your CI service has features that allow you to restrict which pipelines or steps in a pipeline can access the secrets, then ensure that
+  your pipeline/steps can access them.
+* Typo in the name of the environment variables. Refer to our [Cloud Providers]({{< relref "/docs/intro/cloud-providers" >}}) docs and the respective setup pages for each to find the correct
```
Similarly, "Check for typos..."

praneetloke replied:
Done.

cnunciato commented on content/docs/guides/continuous-delivery/troubleshooting-guide.md:
```diff
@@ -0,0 +1,129 @@
+
+### Still need help?
+
+If your error is not related to any of the above categories, it is likely that there is a problem with the code in your infrastructure code.
```
.. or it could be that we just didn't think of it or haven't documented it yet. ;) I do like referring to Slack for this, for sure. I'm not sure I'd point someone into the abyss of the programming model docs to troubleshoot CI problems, though. If they're here, they're likely beyond that level and looking for CI-specific info.

praneetloke replied:
Yes, very much agree with this. Reading this again, it actually doesn't sound nice at all. It seems to blame the user's code with high confidence. I will remove all of this and simply suggest to the user to reach out on our Community Slack.

cnunciato reviewed (commented):
Buncha little things for ya to consider.

praneetloke commented on content/docs/guides/continuous-delivery/troubleshooting-guide.md:
```diff
@@ -0,0 +1,129 @@
+* Ensure that the account represented by the token you are using, has access to the stack.
+  * This can lead to 404s being returned from the Pulumi Service because the token is invalid for any number of reasons.
+* Ensure that you use the fully-qualified stack name when passing the stack name to `pulumi` commands.
+  * A fully-qualified stack name is of the format `<org_name>/<project_name>/<stack_name>`.
```
Yes, very much agree with that.

praneetloke commented on content/docs/guides/continuous-delivery/troubleshooting-guide.md:
```diff
@@ -0,0 +1,129 @@
+  which one you meant based on the project name in your `Pulumi.yaml` file. It makes it explicit for anyone in your team looking at the CI configuration.
+* Ensure that you are running the `pulumi` command from the folder containing the `Pulumi.yaml` file that contains your Pulumi project's name.
+* If your stack has configuration, then the stack configuration file must also be co-located with the project file.
+  * For example, for a stack named `production`, the `Pulumi.production.yaml` file must exist alongside the `Pulumi.yaml`.
```
> Technically it just has to exist at or below the dir holding Pulumi.yaml (e.g., for those who put their stack configs in a subfolder).

Oh! I didn't know that myself! Hmm is there an example of this somewhere?

cnunciato replied:
I don't know of a public example, but the chapter I'm working on covers it, which is the only reason I happened to think of it myself. 😄 For example, if your Pulumi.yaml looked like this:

```
name: some-app
description: Some description
main: program
config: stacks
```

... the corresponding file structure would look like this:

```
$ tree
.
├── Pulumi.yaml
├── program 
│   ├── node_modules
│   │   └── ...
│   ├── tsconfig.json
│   ├── index.ts
│   ├── package-lock.json
│   ├── package.json
│   └── tsconfig.json
└── stacks
    ├── Pulumi.dev.yaml
    └── Pulumi.prod.yaml
```

praneetloke commented on content/docs/guides/continuous-delivery/troubleshooting-guide.md:
```diff
@@ -0,0 +1,129 @@
+
+### Tips
+
+* For `Node` and `Python` runtimes, you will need to run the step to restore the dependencies ahead of time.
```
I will rephrase this sentence.

praneetloke commented on content/docs/guides/continuous-delivery/troubleshooting-guide.md:
```diff
@@ -0,0 +1,129 @@
+* For `.NET` and `Go` runtimes, the dependencies are restored for you automatically when you run `pulumi preview` or `pulumi update --yes`.
+* There is an exception to restoring dependencies automatically for `.NET` when you use a private package feed. You must ensure that the
+package(s) from the private feed are accessible or you can use a pre-built binary (see the `binary` option [here]({{< relref "/docs/intro/concepts/project" >}})) with Pulumi to avoid rebuilding your `.NET` solution again.
+  * Note that if you do choose to use a pre-built binary, automatic plugin acquisition for Pulumi
```
Yeah I just noticed that too myself while I was doing something else and fixed it now.

praneetloke commented on content/docs/guides/continuous-delivery/troubleshooting-guide.md:
```diff
@@ -0,0 +1,129 @@
+* There is an exception to restoring dependencies automatically for `.NET` when you use a private package feed. You must ensure that the
+package(s) from the private feed are accessible or you can use a pre-built binary (see the `binary` option [here]({{< relref "/docs/intro/concepts/project" >}})) with Pulumi to avoid rebuilding your `.NET` solution again.
+  * Note that if you do choose to use a pre-built binary, automatic plugin acquisition for Pulumi
+* You are caching the library dependencies but not the Pulumi plugins. Some services offer dependency caching by capturing a specific folder and restoring
```
It's the latter, but I'll think of a different way to phrase this.

praneetloke commented on content/docs/guides/continuous-delivery/troubleshooting-guide.md:
```diff
@@ -0,0 +1,129 @@
+* You are caching the library dependencies but not the Pulumi plugins. Some services offer dependency caching by capturing a specific folder and restoring
+that folder whenever you run your pipeline subsequently. However, note that Pulumi dependencies have a post-install step that also pulls-down
+a [plugin]({{< relref "/docs/intro/concepts/how-pulumi-works#resource-providers" >}}) binary from our CDN.
+  * So be sure to cache the plugins path as well.
```
Yes, agreed. I have fixed this.

praneetloke commented on content/docs/guides/continuous-delivery/troubleshooting-guide.md:
```diff
@@ -0,0 +1,129 @@
+
+* Cloud provider credentials not specified using a recognized mechanism. In almost all cases, specifying the credentials through environment variables will work.
+  * Some cloud providers also support authentication via their respective CLIs. For example, the Azure provider supports authentication via the Azure CLI.
+  * In the above example, if the Azure CLI is not properly authenticated, then Pulumi will not be able to acquire the necessary credentials.
```
Ah yeah. Fixed it.

praneetloke commented:
I'll go ahead and merge this. I addressed all of the comments. Thank you, @jkisk and @cnunciato for your time reviewing this.  The guide looks much better now!
